E0606: detect validation done in external function

### Bug description

```python
def f(offset):
    validate_offset(offset)

    # compute modifier
    number, suffix = int(offset[:-1]), offset[-1]

    if suffix == "d":
        modifier = f"-{number} days"
    elif suffix == "w":
        modifier = f"-{number * 7} days"
    elif suffix == "y":
        modifier = f"-{number} years"
    # else:
    #    raise ValueError() # This is here to make pylint etc. happy

    # do something with `modifier` here (omitted)
    modifier


def validate_offset(offset):
    if not isinstance(argument, str):
        raise ValueError(f"Invalid offset: {offset!r}")

    suffix = argument[-1:]  # slicing here to handle empty strings
    if suffix not in ("d", "w", "y"):
        raise ValueError(
            f"Invalid offset: {offset!r}\n"
            f"Please specify a string of the format 'X[d/w/y]' "
            "where X is a non-negative integer followed by 'd', 'w', or 'y' "
            "that indicates days, weeks, or years."
        )
```


### Configuration

_No response_

### Command used

```shell
pylint -sn *.py ${SRC_CORE} ${SRC_TEST}
```


### Pylint output

```python
************* Module things.database
things/database.py:1048:42: E0606: Possibly using variable 'modifier' before assignment (possibly-used-before-assignment)
```


### Expected behavior

It is a common pattern for me to validate input in an external validation function. That keeps the code clean and readable.

In this case, I don't expect to have to write the additional code (commented above):

```
    else:
        raise ValueError() # This is here to make pylint etc. happy
```

since the `validate_offset` already raises if modified won't be set.

A potential workaround would be to explicitly mark `modified` will be set using some kind of `asserts` mechanism or comment that exists in the Typescript language (see [here](https://stackoverflow.com/a/71625025)).

### Pylint version

```shell
pylint 3.2.3
astroid 3.2.2
Python 3.11.9
```


### OS / Environment

_No response_

### Additional dependencies

_No response_

--- a/doc/data/messages/p/possibly-used-before-assignment/details.rst
+++ b/doc/data/messages/p/possibly-used-before-assignment/details.rst
@@ -39,7 +39,7 @@ This warns:
         var = 1
 
     if guarded() or other_condition:
-        print(var)
+        print(var)  # [possibly-used-before-assignment]
 
 If you find this surprising, consider that pylint, as a static analysis
 tool, does not know if ``guarded()`` is deterministic or talks to

--- a/doc/data/messages/p/possibly-used-before-assignment/details.rst
+++ b/doc/data/messages/p/possibly-used-before-assignment/details.rst
@@ -1,3 +1,22 @@
+You can use ``assert_never`` to mark exhaustive choices:
+
+.. sourcecode:: python
+
+    from typing import assert_never
+
+    def handle_date_suffix(suffix):
+        if suffix == "d":
+            ...
+        elif suffix == "m":
+            ...
+        elif suffix == "y":
+            ...
+        else:
+            assert_never(suffix)
+
+    if suffix in "dmy":
+        handle_date_suffix(suffix)
+
 If you rely on a pattern like:
 
 .. sourcecode:: python
@@ -5,7 +24,7 @@ If you rely on a pattern like:
     if guarded():
         var = 1
 
-    if guarded():
+    if guarded() or other_condition:
         print(var)  # emits possibly-used-before-assignment
 
 you may be concerned that ``possibly-used-before-assignment`` is not totally useful

--- a/doc/data/messages/p/possibly-used-before-assignment/details.rst
+++ b/doc/data/messages/p/possibly-used-before-assignment/details.rst
@@ -17,7 +17,21 @@ You can use ``assert_never`` to mark exhaustive choices:
     if suffix in "dmy":
         handle_date_suffix(suffix)
 
-If you rely on a pattern like:
+Pylint currently allows repeating the same test like this, even though this
+lets some error cases through, as pylint does not assess the intervening code:
+
+.. sourcecode:: python
+
+    if guarded():
+        var = 1
+
+    # what if code here affects the reuslt of guarded()?
+
+    if guarded():
+        print(var)
+
+But this exception is limited to the repeating the exact same test.
+This warns:
 
 .. sourcecode:: python
 
@@ -25,10 +39,11 @@ If you rely on a pattern like:
         var = 1
 
     if guarded() or other_condition:
-        print(var)  # emits possibly-used-before-assignment
+        print(var)
 
-you may be concerned that ``possibly-used-before-assignment`` is not totally useful
-in this instance. However, consider that pylint, as a static analysis tool, does
-not know if ``guarded()`` is deterministic or talks to
-a database. (Likewise, for ``guarded`` instead of ``guarded()``, any other
-part of your program may have changed its value in the meantime.)
+If you find this surprising, consider that pylint, as a static analysis
+tool, does not know if ``guarded()`` is deterministic or talks to
+a database. For constants (e.g. ``guarded`` versus ``guarded()``),
+this is less of an issue, so in this case,
+``possibly-used-before-assignment`` acts more like a future-proofing style
+preference than an error, per se.

--- a/doc/data/messages/p/possibly-used-before-assignment/details.rst
+++ b/doc/data/messages/p/possibly-used-before-assignment/details.rst
@@ -25,7 +25,7 @@ lets some error cases through, as pylint does not assess the intervening code:
     if guarded():
         var = 1
 
-    # what if code here affects the reuslt of guarded()?
+    # what if code here affects the result of guarded()?
 
     if guarded():
         print(var)

